cmd/clusterctl/client/tree/discovery.go

```diff
@@ -65,10 +67,14 @@ func Discovery(ctx context.Context, c client.Client, namespace, name string, opt
 	}
 
 	// Adds control plane
-	controlPLane, err := external.Get(ctx, c, cluster.Spec.ControlPlaneRef, cluster.Namespace)
-	if err == nil {
-		tree.Add(cluster, controlPLane, ObjectMetaName("ControlPlane"), GroupingObject(true))
+	if cluster.Spec.ControlPlaneRef == nil {
+		return nil, errors.New(fmt.Sprintf("spec.controlPlaneRef for Cluster %v is nil, this is unexpected", client.ObjectKeyFromObject(cluster)))
 	}
+	controlPlane, err := external.Get(ctx, c, cluster.Spec.ControlPlaneRef, cluster.Namespace)
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("failed to get controlplane for cluster %v", client.ObjectKeyFromObject(cluster)))
+	}
+	tree.Add(cluster, controlPlane, ObjectMetaName("ControlPlane"), GroupingObject(true))
```

Comment:
I'm not entirely sure about the history, but could it be that we intentionally didn't error out before because we also support clusters without a controlPlaneRef?

I"m not sure about this, @fabriziopandini please confirm :)

Reply:
Even if a "fully functioning" cluster always has a controlPlaneRef, with ClusterClass we will have intermediate states where the controlPlaneRef will be nil until the cluster is reconciled. I'm not sure  we want to return an error in that case.